addons/viewport/src/preview/withViewport.js

```diff
@@ -9,38 +9,41 @@ import {
 } from '../shared';
 
 function noop() {}
-let handler = noop;
 
-const subscription = () => {
+const subscription = handler => () => {
   const channel = addons.getChannel();
   channel.on(VIEWPORT_CHANGED_EVENT_ID, handler);
   return () => channel.removeListener(VIEWPORT_CHANGED_EVENT_ID, handler);
 };
 
-const setViewport = options => {
+const applyViewportOptions = (options = {}) => {
   const channel = addons.getChannel();
-  handler = options.onViewportChange || noop;
+
+  const handler = options.onViewportChange || noop;
   if (options.onViewportChange) {
-    channel.emit(CoreEvents.REGISTER_SUBSCRIPTION, subscription);
+    channel.emit(CoreEvents.REGISTER_SUBSCRIPTION, subscription(handler));
   }
+
   channel.emit(SET_STORY_DEFAULT_VIEWPORT_EVENT_ID, options.name || DEFAULT_VIEWPORT);
 };
 
-export default function withViewport(nameOrOptions) {
-  const options = typeof nameOrOptions === 'string' ? { name: nameOrOptions } : nameOrOptions;
+const withViewport = makeDecorator({
+  name: 'withViewport',
+  parameterName: 'viewport',
+  wrapper: (getStory, context, { options, parameters }) => {
+    const storyOptions = parameters || options;
+    const viewportOptions =
+      typeof storyOptions === 'string' ? { name: storyOptions } : storyOptions;
+
+    applyViewportOptions(viewportOptions);
```

Comment:
is it needed when there are neither options nor parameters?

Reply:
No, not needed in such case

Reply:
Then you can pass `skipIfNoParametersOrOptions: true` to `makeDecorator`